---
description: コンパイラの警告 (レベル 1) CS4014
title: コンパイラの警告 (レベル 1) CS4014
ms.date: 08/18/2020
f1_keywords:
- CS4014
helpviewer_keywords:
- CS4014
ms.assetid: b9e7c029-eaa9-48d7-8b3c-9823772c9acb
ms.openlocfilehash: 6118ac43672dca053bee57fdf16ac68ad6d61aee
ms.sourcegitcommit: 05d0087dfca85aac9ca2960f86c5efd218bf833f
ms.translationtype: HT
ms.contentlocale: ja-JP
ms.lasthandoff: 03/27/2021
ms.locfileid: "105637547"
---
# <a name="compiler-warning-level-1-cs4014"></a><span data-ttu-id="174f2-103">コンパイラの警告 (レベル 1) CS4014</span><span class="sxs-lookup"><span data-stu-id="174f2-103">Compiler Warning (level 1) CS4014</span></span>

<span data-ttu-id="174f2-104">この呼び出しは待機されなかったため、現在のメソッドの実行は呼び出しの完了を待たずに続行されます。</span><span class="sxs-lookup"><span data-stu-id="174f2-104">Because this call is not awaited, execution of the current method continues before the call is completed.</span></span> <span data-ttu-id="174f2-105">呼び出しの結果に `await` 演算子を適用することを検討してください。</span><span class="sxs-lookup"><span data-stu-id="174f2-105">Consider applying the `await` operator to the result of the call.</span></span>

<span data-ttu-id="174f2-106">現在のメソッドは <xref:System.Threading.Tasks.Task> または <xref:System.Threading.Tasks.Task%601> を返す非同期メソッドを呼び出し、[await](../operators/await.md) 演算子を結果に適用しません。</span><span class="sxs-lookup"><span data-stu-id="174f2-106">The current method calls an async method that returns a <xref:System.Threading.Tasks.Task> or a <xref:System.Threading.Tasks.Task%601> and doesn't apply the [await](../operators/await.md) operator to the result.</span></span> <span data-ttu-id="174f2-107">この非同期メソッドの呼び出しは、非同期タスクを開始します。</span><span class="sxs-lookup"><span data-stu-id="174f2-107">The call to the async method starts an asynchronous task.</span></span> <span data-ttu-id="174f2-108">ただし、 `await` 演算子が適用されないため、プログラムはタスクの完了を待たずに処理を続行します。</span><span class="sxs-lookup"><span data-stu-id="174f2-108">However, because no `await` operator is applied, the program continues without waiting for the task to complete.</span></span> <span data-ttu-id="174f2-109">ほとんどの場合、この動作は想定されるものではありません。</span><span class="sxs-lookup"><span data-stu-id="174f2-109">In most cases, that behavior isn't what you expect.</span></span> <span data-ttu-id="174f2-110">通常は、呼び出し元のメソッドの他の側面が呼び出しの結果に依存します。また、最低でも、呼び出しを含むメソッドから制御が返される前に、呼び出されたメソッドが完了することが想定されます。</span><span class="sxs-lookup"><span data-stu-id="174f2-110">Usually other aspects of the calling method depend on the results of the call or, minimally, the called method is expected to complete before you return from the method that contains the call.</span></span>

<span data-ttu-id="174f2-111">同様に、呼び出された非同期メソッドで発生した例外に対する処理も重要です。</span><span class="sxs-lookup"><span data-stu-id="174f2-111">An equally important issue is what happens to exceptions that are raised in the called async method.</span></span> <span data-ttu-id="174f2-112"><xref:System.Threading.Tasks.Task> または <xref:System.Threading.Tasks.Task%601> を返すメソッド内で発生した例外は、返されたタスクに格納されます。</span><span class="sxs-lookup"><span data-stu-id="174f2-112">An exception that's raised in a method that returns a <xref:System.Threading.Tasks.Task> or  <xref:System.Threading.Tasks.Task%601> is stored in the returned task.</span></span> <span data-ttu-id="174f2-113">このタスクが返されるのを待たない場合や例外を明示的にチェックしない場合、例外は失われます。</span><span class="sxs-lookup"><span data-stu-id="174f2-113">If you don't await the task or explicitly check for exceptions, the exception is lost.</span></span> <span data-ttu-id="174f2-114">このタスクが返されるのを待機する場合は、例外が再スローされます。</span><span class="sxs-lookup"><span data-stu-id="174f2-114">If you await the task, its exception is rethrown.</span></span>

<span data-ttu-id="174f2-115">ベスト プラクティスとしては、常に呼び出しを待機する必要があります。</span><span class="sxs-lookup"><span data-stu-id="174f2-115">As a best practice, you should always await the call.</span></span>

<span data-ttu-id="174f2-116">非同期呼び出しの完了を待つ必要がなく、呼び出されたメソッドで例外が発生しないことが確実である場合に限り、警告を抑制することを検討してください。</span><span class="sxs-lookup"><span data-stu-id="174f2-116">You should consider suppressing the warning only if you're sure that you don't want to wait for the asynchronous call to complete and that the called method won't raise any exceptions.</span></span> <span data-ttu-id="174f2-117">その場合は、呼び出しのタスクの結果を変数に割り当てることで警告を抑制することができます。</span><span class="sxs-lookup"><span data-stu-id="174f2-117">In that case, you can suppress the warning by assigning the task result of the call to a variable.</span></span>

<span data-ttu-id="174f2-118">次の例で、警告を発生させる方法、警告を抑制する方法、呼び出しを待機する方法を示します。</span><span class="sxs-lookup"><span data-stu-id="174f2-118">The following example shows how to cause the warning, how to suppress it, and how to await the call.</span></span>

```csharp
static async Task CallingMethodAsync(int millisecondsDelay)
{
    Console.WriteLine("  Entering calling method.");

    // Call #1.
    // Call an async method. Because you don't await it, its completion
    // isn't coordinated with the current method, CallingMethodAsync.
    // The following line causes warning CS4014.
    CalledMethodAsync(millisecondsDelay);

    // Call #2.
    // To suppress the warning without awaiting, you can assign the
    // returned task to a variable. The assignment doesn't change how
    // the program runs. However, recommended practice is always to
    // await a call to an async method.

    // Replace Call #1 with the following line.
    // Task delayTask = CalledMethodAsync(millisecondsDelay);

    // Call #3
    // To contrast with an awaited call, replace the unawaited call
    // (Call #1 or Call #2) with the following awaited call. Best
    // practice is to await the call.

    // await CalledMethodAsync(millisecondsDelay);

    Console.WriteLine("  Returning from calling method.");
}

static async Task CalledMethodAsync(int millisecondsDelay)
{
    Console.WriteLine("    Entering called method, starting and awaiting Task.Delay.");

    await Task.Delay(millisecondsDelay);

    Console.WriteLine("    Task.Delay is finished--returning from called method.");
}
```

<span data-ttu-id="174f2-119">この例では、呼び出し 1 または呼び出し 2 を選択した場合、完了するまで待機されない非同期メソッド `CalledMethodAsync` は、呼び出し元 `CallingMethodAsync` と呼び出し元の呼び出し元の両方が完了した後に完了します。</span><span class="sxs-lookup"><span data-stu-id="174f2-119">In the example, if you choose Call #1 or Call #2, the unawaited async method `CalledMethodAsync` finishes after both its caller `CallingMethodAsync` and the caller's caller is complete.</span></span> <span data-ttu-id="174f2-120">次の出力の最後の行に、呼び出されたメソッドがいつ完了したかが示されています。</span><span class="sxs-lookup"><span data-stu-id="174f2-120">The last line in the following output shows you when the called method finishes.</span></span> <span data-ttu-id="174f2-121">この出力には、完全な例の `CallingMethodAsync` を呼び出すイベント ハンドラーへのエントリとその終了が示されています。</span><span class="sxs-lookup"><span data-stu-id="174f2-121">Entry to and exit from the event handler that calls `CallingMethodAsync` in the full example are marked in the output.</span></span>

```console
Entering the Click event handler.
  Entering calling method.
    Entering called method, starting and awaiting Task.Delay.
  Returning from calling method.
Exiting the Click event handler.
    Task.Delay is finished--returning from called method.
```

<span data-ttu-id="174f2-122">また、[#pragma warning](../preprocessor-directives.md#pragma-warning) ディレクティブを使用してコンパイラの警告を抑制することもできます。</span><span class="sxs-lookup"><span data-stu-id="174f2-122">You can also suppress compiler warnings by using [#pragma warning](../preprocessor-directives.md#pragma-warning) directives.</span></span>

## <a name="example"></a><span data-ttu-id="174f2-123">例</span><span class="sxs-lookup"><span data-stu-id="174f2-123">Example</span></span>

<span data-ttu-id="174f2-124">次のコンソール アプリケーションには、前の例のメソッドが含まれています。</span><span class="sxs-lookup"><span data-stu-id="174f2-124">The following console application contains the methods from the previous example.</span></span> <span data-ttu-id="174f2-125">このアプリケーションを設定するには、次の手順を実行します。</span><span class="sxs-lookup"><span data-stu-id="174f2-125">The following steps set up the application.</span></span>

1. <span data-ttu-id="174f2-126">コンソール アプリケーションを作成し、それに `AsyncWarning` という名前を付けます。</span><span class="sxs-lookup"><span data-stu-id="174f2-126">Create a console application, and name it `AsyncWarning`.</span></span>
1. <span data-ttu-id="174f2-127">Visual Studio Code エディターで、*Program.cs* ファイルを選択します。</span><span class="sxs-lookup"><span data-stu-id="174f2-127">In the Visual Studio Code Editor, choose the *Program.cs* file.</span></span>
1. <span data-ttu-id="174f2-128">*Program.cs* 内のコードを、次のコードに置き換えます。</span><span class="sxs-lookup"><span data-stu-id="174f2-128">Replace the code in *Program.cs* with the following code.</span></span>

    ```csharp
    using System;
    using System.Threading.Tasks;

    namespace AsyncWarning
    {
        class Program
        {
            static async Task Main()
            {
                Console.WriteLine("Entering Main() application entry point.");

                int millisecondsDelay = 2000;
                await CallingMethodAsync(millisecondsDelay);

                Console.WriteLine("Exiting Main() application entry point.");

                await Task.Delay(millisecondsDelay + 500);
            }

            static async Task CallingMethodAsync(int millisecondsDelay)
            {
                Console.WriteLine("  Entering calling method.");

                // Call #1.
                // Call an async method. Because you don't await it, its completion
                // isn't coordinated with the current method, CallingMethodAsync.
                // The following line causes warning CS4014.
                // CalledMethodAsync(millisecondsDelay);

                // Call #2.
                // To suppress the warning without awaiting, you can assign the
                // returned task to a variable. The assignment doesn't change how
                // the program runs. However, recommended practice is always to
                // await a call to an async method.

                // Replace Call #1 with the following line.
                //Task delayTask = CalledMethodAsync(millisecondsDelay);

                // Call #3
                // To contrast with an awaited call, replace the unawaited call
                // (Call #1 or Call #2) with the following awaited call. Best
                // practice is to await the call.

                // await CalledMethodAsync(millisecondsDelay);

                Console.WriteLine("  Returning from calling method.");
            }

            static async Task CalledMethodAsync(int millisecondsDelay)
            {
                Console.WriteLine("    Entering called method, starting and awaiting Task.Delay.");

                await Task.Delay(millisecondsDelay);

                Console.WriteLine("    Task.Delay is finished--returning from called method.");
            }
        }

        // Output with Call #1 or Call #2. (Wait for the last line to appear.)

        // Entering Main() application entry point.
        //   Entering calling method.
        //     Entering called method, starting and awaiting Task.Delay.
        //   Returning from calling method.
        // Exiting Main() application entry point.
        //     Task.Delay is finished--returning from called method.

        // Output with Call #3, which awaits the call to CalledMethodAsync.

        // Entering Main() application entry point.
        //   Entering calling method.
        //     Entering called method, starting and awaiting Task.Delay.
        //     Task.Delay is finished--returning from called method.
        //   Returning from calling method.
        // Exiting Main() application entry point.
    }
    ```

1. <span data-ttu-id="174f2-129"><kbd>F5</kbd> キーを選択し、プログラムを実行します。</span><span class="sxs-lookup"><span data-stu-id="174f2-129">Select the <kbd>F5</kbd> key to run the program.</span></span>

<span data-ttu-id="174f2-130">想定される出力がコードの最後に表示されます。</span><span class="sxs-lookup"><span data-stu-id="174f2-130">The expected output appears at the end of the code.</span></span>

## <a name="see-also"></a><span data-ttu-id="174f2-131">関連項目</span><span class="sxs-lookup"><span data-stu-id="174f2-131">See also</span></span>

- [<span data-ttu-id="174f2-132">await</span><span class="sxs-lookup"><span data-stu-id="174f2-132">await</span></span>](../operators/await.md)
- [<span data-ttu-id="174f2-133">async および await を使用した非同期プログラミング</span><span class="sxs-lookup"><span data-stu-id="174f2-133">Asynchronous programming with async and await</span></span>](../../programming-guide/concepts/async/index.md)
